### PR TITLE
Put opening times as params

### PR DESCRIPTION
**Pull request by BlairArchibald**
Finally got around to putting the opening times as parameters (as part of https://github.com/ScottishCovidResponse/SCRCIssueTracking/issues/410) which gives us some more flexibility. For example, school times were 9-5, but we might want to make some 9-15 (or similar) to better model Primary school times. This makes it much easy to try this sort of thing.

I've made it recreate the existing functionality which leaves us with a little weirdness in that times/shifts are *either* assigned probabilistically, or based on a size (smaller shops open less). This feels odd, but I think we may as well wait for better data before we decide which way to go.

### CODE REVIEW

**Comment by BlairArchibald** [comment]
I'm happy to ignore the Cadacy issues - I think for lambda function based static fields we can treat these as methods and not have them next to the other data fields.

File: src/main/java/uk/co/ramp/covid/simulation/parameters/BuildingProperties.java [hunk 1]
```diff
@@ -33,16 +33,32 @@
 
     /** Dates when schools are closed */
     public List<DateRange> schoolHolidays = null;
-    
+
+    /** Openning times/Shifts */
```
Comment (bobtaylor12) [inline src/main/java/uk/co/ramp/covid/simulation/parameters/BuildingProperties.java:37]: Typo

File: src/main/java/uk/co/ramp/covid/simulation/parameters/BuildingProperties.java [hunk 2]
```diff
@@ -33,16 +33,32 @@
...
+    public List<BuildingTimeParameters> careHomeTimes;
+    public List<BuildingTimeParameters> restaurantTimes;
+    public List<BuildingTimeParameters> shopTimes;
+    public List<BuildingTimeParameters> schoolTimes;
+    public List<BuildingTimeParameters> officeTimes;
+    public List<BuildingTimeParameters> constructionSiteTimes;
+    public List<BuildingTimeParameters> hospitalTimes;
+
```
Comment (bobtaylor12) [inline src/main/java/uk/co/ramp/covid/simulation/parameters/BuildingProperties.java:45]: Should nurseryTimes be declared here?

File: parameters/example_population_params.json [hunk 3]
```diff
@@ -207,7 +207,251 @@
...
+                  "start": 8,
+                  "end": 15,
+                  "days": [3,4,5,6]
+                },
+                {
+                  "start": 15,
+                  "end": 22,
+                  "days": [3,4,5,6]
+                }]
+            }
```
Comment (bobtaylor12) [inline parameters/example_population_params.json:453]: What about nurseryTimes?

File: src/main/java/uk/co/ramp/covid/simulation/parameters/BuildingProperties.java [hunk 4]
```diff
@@ -33,16 +33,32 @@
...
+                && schoolExpectedInteractionsPerHour >= 0
+                && shopExpectedInteractionsPerHour >= 0
+                && careHomeExpectedInteractionsPerHour >= 0
+                && careHomeTimes.stream().allMatch(BuildingTimeParameters::isValid)
+                && restaurantTimes.stream().allMatch(BuildingTimeParameters::isValid)
+                && shopTimes.stream().allMatch(BuildingTimeParameters::isValid)
+                && schoolTimes.stream().allMatch(BuildingTimeParameters::isValid)
+                && officeTimes.stream().allMatch(BuildingTimeParameters::isValid)
+                && constructionSiteTimes.stream().allMatch(BuildingTimeParameters::isValid)
+                && careHomeTimes.stream().allMatch(BuildingTimeParameters::isValid);
```
Comment (bobtaylor12) [inline src/main/java/uk/co/ramp/covid/simulation/parameters/BuildingProperties.java:61]: Add nurseryTimes?

File: src/main/java/uk/co/ramp/covid/simulation/place/CommunalPlace.java [hunk 5]
```diff
@@ -217,4 +216,55 @@ public void open() {
...
+        return shifts.getNext();
+    }
+
+    public final boolean isFullyStaffed() {
+        return nStaff >= shifts.size();
+    }
+
+    protected abstract List<BuildingTimeParameters> getTimeInfo();
+
+    // setTimes configures openning times and shifts
```
Comment (bobtaylor12) [inline src/main/java/uk/co/ramp/covid/simulation/place/CommunalPlace.java:231]: Typo - "opening"

File: src/main/java/uk/co/ramp/covid/simulation/place/Nursery.java [hunk 6]
```diff
@@ -29,13 +30,7 @@ public void reportInfection(Time t, Person p, DailyStats s) {
...
-    public boolean isFullyStaffed() {
-        return nStaff > 0;
-    }
-
-    @Override
-    public Shifts getShifts() {
-        nStaff++;
-        return Shifts.schoolTimes();
+    protected List<BuildingTimeParameters> getTimeInfo() {
+        return PopulationParameters.get().buildingProperties.schoolTimes;
```
Comment (bobtaylor12) [inline src/main/java/uk/co/ramp/covid/simulation/place/Nursery.java:34]: I would argue that school opening times and nursery opening times may be different.

File: src/main/java/uk/co/ramp/covid/simulation/population/Shifts.java [hunk 7]
```diff
@@ -62,4 +77,49 @@ public static Shifts schoolTimes() {
...
+    public static final JsonSerializer<Shifts> serializer = (src, typeOfSrc, context) -> {
+        JsonObject o = new JsonObject();
+        List<Integer> days = new ArrayList<>();
+        src.shifts.forEach((k,v) -> {
+            if (v != null) {
+                days.add(k);
+            }
+        });
+
+        // Shift infomration is shared over days so we just take the first
```
Comment (bobtaylor12) [inline src/main/java/uk/co/ramp/covid/simulation/population/Shifts.java:99]: typo

**Comment by BlairArchibald** [comment]
Hi @bobtaylor12  - totally agree nursery time should be settable. This should now be fixed! Thanks for the review!

**Pull request merged**